api/v1alpha3/types.go

```diff
@@ -263,4 +263,6 @@ type Bastion struct {
 	FloatingIP string `json:"floatingIP,omitempty"`
 	//+optional
 	SecurityGroups []SecurityGroupParam `json:"securityGroups,omitempty"`
+	//+optinonal
+	FailureDomain string `json:"failureDomain,omitempty"`
```

Comment:
Is this really a failuredomain in the generic sense that we have it in the KubeadmControlPlane?

I would have guessed we have only one bastion host and we simply specific the availabilty zone here in which it should be created.

Reply:
Looks to me like you're right. But when creating an Instance (like both Bastion and control plane nodes seem to be doing similarly), they just translate the FailureDomain from the CAPx spec into OpenStack AZs here?

Example:
https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/30ddae05f68056feb692318bb8c3479869eee034/pkg/cloud/services/compute/instance.go#L228

So I guess this looks fine to me. 👍

Reply:
I'm not entirely sure but I think we only do this for control plane machines. For all other machines it's just setting the az on a OpenStackMachineSpec

FailureDomain as a concept is meant to abstract away from Specific IaaS provider because failure domains can be very differently (especially regarding hierarchy) between different cloud providers.

Reply:
Just forget everything I've said :)

You're absolutely right, we're just working with FailureDomains everywhere. Not sure how I got the idea that we specify az's on any resources. Probably mixed it up with our internal proprietar non-ClusterAPI implementation :)

Reply:
Okay so overall looks good to me. Although I suspect before going to beta someday we should do a full API redesign (not especially because of this PR but our resources could use a well thought out design after the history we have behind us :)).

Reply:
Well, I guess you're right too? 😊 

`OpenStackCluster` has `spec.controlPlaneAvailabilityZones`
`MachineDeployment` has `spec.template.spec.failureDomain`

Reply:
We should probably start an issue with API inconsistencies which we can then all fix when we redesign it. I guess for v1alpha4 it's not worth the breaking change

Reply:
@jichenjc When I borrow bastion concept from CAPA, I too wondered why bastion does not use AWSMachineSpec. 
Bastion is not openStackMachine, but member of OpenStackClusterSpec. So I thought we should not use OpenStackMachineSpec. 
FailureDomain is somehow considered related to Cluster API. Bastion concept is not directly related to Cluster API. 

Simply I suggest to add `AvailabilityZone string`

Reply:
I understand , but I think we don't need fully follow AWS provider here? 
bastion to me it's `some special param` + `a VM` , so if we keep adding stuffs like root volume, security groups 
ssh keys, then finally it will became a MachineSpec any way..

I can move to AvailabilityZone